Add path exclusion to mTLS and BasicAuth authentication

rzetelskik opened the pull request:
This PR introduces support for path exclusion from authentication for both mTLS and BasicAuth.
The main issue here was allowing for excluding the requests, which request a URL matching an excepted path, from
mTLS authentication. This PR achieves this by lowering the requirements in TLSConfig and delegating the certificate verification to a dedicated middleware - an approach heavily based on how [kubernetes/apiserver](https://github.com/kubernetes/apiserver) handles authentication and authorization. To match this approach and allow for path exclusion for both types of authentication, basic authentication is also moved to a dedicated middleware.

As discussed in https://github.com/prometheus/prometheus/issues/9166, the need for this comes from kubelet, the Kubernetes node agent, not allowing the use of certificates or safely providing credentials for liveness/readiness probes. See also https://github.com/prometheus-operator/prometheus-operator/issues/5419.

This PR also takes into account the comments in the previous attempts to add support for this: https://github.com/prometheus/exporter-toolkit/pull/106 and https://github.com/prometheus/exporter-toolkit/pull/70.

Fixes https://github.com/prometheus/prometheus/issues/9166

Edit:
Some implementation details, as requested by @bboreham :
This PR introduces an `Authenticator` interface.
```go
type Authenticator interface {
	Authenticate(*http.Request) (bool, string, error)
}
```
The `Authenticate` method takes an `*http.Request` and decides whether the given request passes authentication.
The `func WithAuthentication(handler http.Handler, authenticator Authenticator, logger log.Logger) http.Handler` middleware is also introduced. It authenticates every request and passes it to the next handler on success.

The `ChainAuthenticator` is also implemented to conjunct multiple Authenticators. Its implementation of `Authenticate` fails to authenticate the request unless it authenticates against all chained Authenticators successfully. 

There are two instantiations of the interface implemented in this PR: `BasicAuthAuthenticator` and `X509Authenticator`.

`BasicAuthAuthenticator` isolates all the code responsible for basic auth authentication that was there before in `webHandler` without any changes to the logic.

`X509Authenticator` implements the part of mTLS server handshake that is normally executed for connections with `ClientAuth` higher than `VerifyClientCertIfGiven` (https://cs.opensource.google/go/go/+/master:src/crypto/tls/handshake_server.go;l=878).
The crux of this PR is the following: one can't just exclude certain URLs from mTLS authentication. There's simply no access to this information at that level of encapsulation. Therefore `ClientAuth` in tlsConfig is always capped at `RequestClientCert`, and the actual certificate verification is delegated to the x509 middleware (https://github.com/prometheus/exporter-toolkit/pull/151/commits/ab8e6f53209253c5d7117b1797523119afc3a01f#diff-ee4d3f619ee85cfe4114c969b9777f1936391e90d3198df35fc6f0145077badeR246).

For the actual path exclusion an `Exceptor` interface is introduced.
```go
type Exceptor interface {
	IsExcepted(r *http.Request) bool
}
```
`func WithExceptor(handler http.Handler, authenticator Authenticator, exceptor Exceptor, logger log.Logger) http.Handler` middleware calls `IsExcepted` and checks, based on a requested URL, whether a request is excepted from authentication. If so, it passes it directly to the provided handler, otherwise the handler is preceded with the authenticator.

rzetelskik committed:
Add path exclusion support to BasicAuth authentication

Signed-off-by: Kacper Rzetelski <kacper.rzetelski@scylladb.com>

rzetelskik committed:
Add path exclusion support to mTLS authentication

Signed-off-by: Kacper Rzetelski <kacper.rzetelski@scylladb.com>

rzetelskik commented:
cc @roidelapluie @SuperQ

rzetelskik commented:
@roidelapluie @SuperQ any chance to have someone review it?

bboreham commented:
I don't know much about mTLS, but the title and description seem way too small for a +1200-line change.

bboreham commented:
> The ChainAuthenticator is also implemented to conjunct multiple Authenticators.

Why?  Maybe you are doing a large-scale refactoring?  Please explain.

It seems possible that this line should go higher up in the explanation:

> The crux of this PR is the following: one can't just exclude certain URLs from mTLS authentication.

rzetelskik commented:
>It seems possible that this line should go higher up in the explanation:
>>The crux of this PR is the following: one can't just exclude certain URLs from mTLS authentication.

Good point. I changed the main part of the description to include this.

>>The ChainAuthenticator is also implemented to conjunct multiple Authenticators.

>Why?

I'm not sure I understand what your question is here. Why is the `ChainAuthenticator` introduced? I believe it's just cleaner than conditionally nesting the handlers with authenticators. It also allows for adding path exception to a collection of authenticators at once.

> Maybe you are doing a large-scale refactoring? Please explain.

I don't think I'm doing much refactoring in this PR besides moving the code responsible for basic auth authentication to a middleware, but I've explained above the reasons for doing so. Other than that I've tried to keep the changes in the existing code minimal, so I don't understand where this is coming from.

bboreham commented:
> I don't think I'm doing much refactoring in this PR
> only about 200-300 LoC

OK, we have a different view on that.  The whole project is currently 1823 lines.

Thank you, the rationale to create a middleware and chain makes sense, it just wasn't in the description first time around.

rzetelskik commented:
ping @SuperQ @roidelapluie

rzetelskik commented:
ping @SuperQ @roidelapluie

roidelapluie commented:
I do  not feel confident writing TLS code ourselves is benefitting for the project.

rzetelskik commented:
First and foremost, I don't agree that this approach includes "writing TLS code" at all. As explained in the PR description, all it does is it lowers certificate requirements on handshake and delegates the required certificate verification to a middleware. And for the verification itself it uses functions from a standard library, which is nowhere near to actually implementing the protocol. And even the code used to verify the certificates comes from the stdlib implementation: https://cs.opensource.google/go/go/+/master:src/crypto/tls/handshake_server.go;l=858.

Having said that, I'd be happy to hear your suggestions on how to implement the path exclusion for mTLS in a different manner, without "writing TLS code ourselves", as I believe you @roidelapluie were the one to suggest extending path exclusion to mTLS in one of the previous attempts to do so: https://github.com/prometheus/exporter-toolkit/pull/106#issuecomment-1246477333.

rzetelskik commented:
@roidelapluie ping

rzetelskik commented:
@roidelapluie ping

ashish1099 commented:
any update on this one ?

rzetelskik commented:
> any update on this one ?

My team still needs this for prometheus-operator and I'm happy to work on merging this. As you can see, the maintainers don't seem particularly interested though.

francesco-furnari commented:
any update  ?

bboreham commented:
I have added this for discussion at a future Prometheus Dev Summit
https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit

rzetelskik commented:
Thanks @bboreham!
Just to comment on the entry in the doc - describing this PR as "refactoring of mTLS code" doesn't do it justice imho. It adds a feature as described in [1], and it does so in a way proposed by one of the Prometheus team members in a comment on one of the earlier PRs trying to solve this [2]. I asked for confirmation if this would be the desired approach but didn't get any responses then.

The use cases behind it: [3], [4].

[1] - https://github.com/prometheus/prometheus/issues/9166
[2] - https://github.com/prometheus/exporter-toolkit/pull/106#issuecomment-1246477333
[3] - https://github.com/prometheus-operator/prometheus-operator/issues/5419
[4] - https://github.com/prometheus-operator/prometheus-operator/issues/4200

matthiasr commented:
Hey, sorry for the delay in getting around to this. As per the Dev Summit, I am picking up the review hat for this. Please bear with me as I work my way through it 😄

matthiasr reviewed (commented):
Sorry for the long delay, it did take a while to get through everything with the care it deserves. Overall this looks good! Thank you for adding all the tests 🙏 they really helped me understand and trust what is going on.

My main ask is documentation, documentation, documentation. Both for the end user, at the "how and why to configure my exporter" level, and for users of the library. If we export it, someone will use it, and they need to know the semantics and what exactly happens when. This is particularly dicy around the TLS options – with no documentation in the toolkit to get them started, we cannot expect developers who use this to jump through and parse all the details of the Go TLS configuration options we are now exposing them to. In general, if we don't _need_ to export something, I would err on the side of keeping it private – it's easy to later make things public when there's a clear use case, but hard to keep track of the big library surface we are opening up here.

Could you please rebase the PR?

rzetelskik commented on web/authentication/exceptor_test.go:
```diff
@@ -0,0 +1,119 @@
+
+	for _, tt := range ts {
+		t.Run(tt.Name, func(t *testing.T) {
+			tt := tt
```
This was necessary before 1.22 https://go.dev/blog/loopvar-preview, I'll get rid of it

rzetelskik commented:
@matthiasr thanks for the detailed review! I'll try to get to rebasing it and going through your comments in the following weeks, bear with me as it's been quite a while since it submitted the PR, so I'll probably have to reiterate on some of the choices I made.

rzetelskik commented on web/authentication/x509/x509.go:
```diff
@@ -0,0 +1,117 @@
+	"github.com/prometheus/exporter-toolkit/web/authentication"
+)
+
+type RequireClientCertsFunc func() bool
+
+type VerifyOptionsFunc func() x509.VerifyOptions
+
+type VerifyPeerCertificateFunc func([][]byte, [][]*x509.Certificate) error
```
In general, by documenting this do you mean adding comments for the functions/interfaces you specifically pointed to, or rather adding a page/section to `docs/`? Or both?

rzetelskik replied:
I removed the types for these particular functions in the process but I added comments to the structure and Authenticate func to describe the intent and reference the counterparts from crypto/tls that this is implementation is mimicking. I believe it will also make it easier to keep it up to date.

rzetelskik commented on web/authentication/exceptor.go:
```diff
@@ -0,0 +1,65 @@
+	}
+}
+
+var _ Exceptor = &PathExceptor{}
```
It's an interface guard meant to statically verify that `PathExceptor` conforms to `Exceptor` interface at compile time. It's not exactly necessary in this particular case, but at the very least it's helpful for developers. It's a known pattern [1] so I don't really see the point of documenting it.

[1] - https://go.dev/doc/effective_go#blank_implements (see `Interface checks`)

rzetelskik commented on web/authentication/x509/x509.go:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	var verifiedChains [][]*x509.Certificate
+	if len(r.TLS.PeerCertificates) > 0 && x.verifyOptionsFn != nil {
```
we don't return early if `x.requireClientCertsFn()` returns false

rzetelskik commented on web/authentication/exceptor_test.go:
```diff
@@ -0,0 +1,119 @@
+			ExcludedPaths:    []string{"/somepath"},
+			URI:              "/somepath",
+			ExpectedExcepted: true,
+		},
```
Imo `/somepath` shouldn't be considered a prefix of `/somepath/someotherpath` in either case here as it's not a subpath, but overall specifying whether subpaths should be excluded is a good point and I added a test for now expecting `/somepath` to not be excluded when `/` is excluded.

If you see a use case behind excluding prefixes I can add it, although imo it would have to only apply to subpaths and be very explicit to avoid potential errors from the user side. Perhaps explicitly specifying an `Exact` or `Prefix` kind, like [k8s ingress does](https://kubernetes.io/docs/concepts/services-networking/ingress/#path-types)? Let me know what you think, personally I'm fine with only allowing exact paths.

matthiasr replied:
No need to add support for excluding prefixes, I just wanted to make sure we are explicit about the semantics and have a test to enforce them – what I want to avoid is that we _accidentally_ turn this into a prefix-match at some point in the future. I see that there's a test for this now, so I'm happy 😄

rzetelskik commented on web/authentication/exceptor_test.go:
```diff
@@ -0,0 +1,119 @@
+			}
+
+			if !tt.ExpectedExcepted && excepted {
+				t.Fatalf("Expected path to not be excepted, it was")
```
done

rzetelskik commented on docs/web-configuration.md:
```diff
@@ -107,6 +107,10 @@ http_server_config:
+  
+# A list of HTTP paths to be excepted from authentication.
+auth_excluded_paths:
+[ - <string> ]
```
With Prometheus Operator, the probes are set up by the controllers. So in this case it should also be the operator that configures the excluded paths for probes, not the direct users of the API.

rzetelskik commented on web/authentication/chain/chain.go:
```diff
@@ -0,0 +1,51 @@
+			return false, reason, nil
+		}
+
+		if len(reason) > 0 {
+			reasons = append(reasons, reason)
+		}
```
> we only return this one authenticator's reason when it does deny

Yes, both BasicAuth and X509 only return a reason when they deny, but at the moment nothing specifically prohibits them from returning a reason when they accept.

> What is the use case for reasons on success?

I don't really see one.

> Could you add a test that illustrates this behavior?

Before I do that, what would be the desired behaviour in your opinion? Should we just say that a reason is undefined when an Authenticator accepts? In which case the Chain authenticator could also just ignore it?

rzetelskik commented on web/authentication/basicauth/basicauth.go:
```diff
@@ -0,0 +1,84 @@
+
+	if !auth {
+		return false, "No credentials in request", nil
+	}
```
Right, looks like I removed `WWW-Authenticate: Basic` header. I'll have to think about how to handle it with authenticators.

As for adding the realm though, I think it's out of scope for this PR - I wouldn't like it to change the current behaviour.
Also it looks like it's not necessarily required any more: https://datatracker.ietf.org/doc/html/rfc7235#appendix-A?

rzetelskik replied:
I added a possibility for an authenticator to optionally return a `HTTPChallenge` struct which can then be used to add headers by the middleware

matthiasr replied:
OK cool, if `realm` is not required then we don't need to add that complexity!

rzetelskik commented on web/authentication/x509/x509.go:
```diff
@@ -0,0 +1,117 @@
+
+// DefaultVerifyOptions returns VerifyOptions that use the system root certificates, current time,
+// and requires certificates to be valid for client auth (x509.ExtKeyUsageClientAuth)
+func DefaultVerifyOptions() x509.VerifyOptions {
```
> do we need to export this?

We don't, I made it a private helper of the unit tests. It's no longer bound to how we set it in tls-config.

> how can we rename it to make it clearer that it's not to be used by itself? Maybe baseVerifyOptions?

renamed, thanks!

rzetelskik commented on web/authentication/chain/chain.go:
```diff
@@ -0,0 +1,51 @@
+
+type ChainAuthenticator []authentication.Authenticator
+
+func (c ChainAuthenticator) Authenticate(r *http.Request) (bool, string, error) {
```
added comments

rzetelskik commented on web/authentication/exceptor.go:
```diff
@@ -0,0 +1,65 @@
+	"github.com/go-kit/log/level"
+)
+
+type Exceptor interface {
```
added comments

rzetelskik commented on web/authentication/exceptor.go:
```diff
@@ -0,0 +1,65 @@
+	return p.excludedPaths[r.URL.Path]
+}
+
+func NewPathExceptor(excludedPaths []string) Exceptor {
```
added comments

rzetelskik commented on web/authentication/x509/x509.go:
```diff
@@ -0,0 +1,117 @@
+	return true, "", nil
+}
+
+func NewX509Authenticator(requireClientCertsFn RequireClientCertsFunc, verifyOptionsFn VerifyOptionsFunc, verifyPeerCertificateFn VerifyPeerCertificateFunc) authentication.Authenticator {
```
Added comments in the structure.

matthiasr reviewed (approved):
> As for exporting the authentication package - perhaps we could just move it to internal/?

Oh yeah that's a good idea! Let's make anything that's not supposed to be used outside of the exporter-toolkit itself internal. Other than that, this looks ready to go for me!

rzetelskik commented:
@matthiasr thanks! I moved it to `/web/internal/authentication`.

matthiasr reviewed (approved):
@SuperQ this is ready to go!

matthewbjones commented:
This is great, as Kamal (https://kamal-deploy.org/) uses `kamal-proxy` and that relies upon a healthcheck endpoint and does not currently support HTTP Basic Auth, so in order to deploy Prometheus using the Kamal ecosystem, it requires that the healthcheck endpoint served by Prometheus to be exempt from the HTTP Basic Auth.

rzetelskik commented:
@SuperQ is there anything you need from my side to merge this one?

SuperQ reviewed (commented):
One question, what about including additional paths in code?

For example, if I'm implementing an exporter with a health endpoint that I want to bypass auth, wouldn't I want to be able to do that by default so that users don't have to manually modify their local configs?

rzetelskik commented:
> One question, what about including additional paths in code?
> 
> For example, if I'm implementing an exporter with a health endpoint that I want to bypass auth, wouldn't I want to be able to do that by default so that users don't have to manually modify their local configs?

@SuperQ what would be a good place to to do this in your opinion? Without modifying the signature of [`Serve`](https://github.com/prometheus/exporter-toolkit/blob/03bc5df704d861a6606860bff992be0ee962c04e/web/tls_config.go#L346), I could wire it into [`FlagConfig`](https://github.com/prometheus/exporter-toolkit/blob/03bc5df704d861a6606860bff992be0ee962c04e/web/tls_config.go#L64). Would something like the below work?
```
type FlagConfig struct {
        ...
        
	AdditionalAuthExcludedPaths *[]string
}
```

SuperQ commented:
`FlagConfig` seems like a reasonable idea to me.

What do you think @matthiasr?

matthiasr commented:
Hmm, it seems odd to me to add something to FlagConfig that's not, you know, a flag. If we add it to that, we should go the whole mile and _expose_ this as a flag.

Alternatively, if you want it to be configurable on Serve without forcing everyone to change their invocations, you could add it using a [functional options pattern](https://www.sohamkamani.com/golang/options-pattern/); since this uses a vararg `...opts` it would be valid to pass zero options. We can decide later if any of the current arguments on Serve should be optional in a breaking change.

matthiasr commented:
I would also argue that this PR is already very large and complex, and because of that has been difficult to get merged. Adding support for optional paths defined in exporter code seems like a well-contained, additional feature that we could add in a separate PR?

rzetelskik commented:
> I would also argue that this PR is already very large and complex, and because of that has been difficult to get merged. Adding support for default paths defined in exporter code seems like a well-contained, additional feature that we could add in a separate PR?

Agreed, it seems to be an extension that can well enough go separately, in which case the approach most suitable for working with exporters directly could be discussed in a dedicated issue.

Chadley1993 commented:
Is there any ETA on this merge?

Chadley1993 commented:
Is anyone still looking into this?

rzetelskik commented:
No response from maintainers in over two years.
